### PR TITLE
[Merged by Bors] - feat(NumberTheory/LegendreSymbol): `@[csimp]` rule for `jacobiSym` and `legendreSym`

### PR DESCRIPTION
**Pull request by hanwenzhu**
Main definitions:
* `def jacobiSym.fastJacobiSym (a : ℤ) (b : ℕ) : ℤ`
* `@[csimp] theorem jacobiSym.fastJacobiSym.eq : jacobiSym = jacobiSym.fastJacobiSym` 
* `def legendreSym.fastLegendreSym (p : ℕ) [Fact p.Prime] (a : ℤ) : ℤ`
* `@[csimp] theorem legendreSym.fastLegendreSym.eq : legendreSym = legendreSym.fastLegendreSym`

Also added tests

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LegendreSymbol/JacobiSymbol.lean [hunk 1]
```diff
@@ -501,3 +503,160 @@ theorem mod_right (a : ℤ) {b : ℕ} (hb : Odd b) : J(a | b) = J(a | b % (4 * a
 end jacobiSym
 
 end Jacobi
+
+
+section FastJacobi
+
+/-!
+### Fast computation of the Jacobi symbol
+We follow the implementation as in `Mathlib.Tactic.NormNum.LegendreSymbol`.
```
Comment (urkud) [inline Mathlib/NumberTheory/LegendreSymbol/JacobiSymbol.lean:512]: Please describe the algorithm in a few sentences here or in the module docstring. Also, is it possible to deduplicate the code between this file and the `NormNum` extension?

Comment (hanwenzhu) [reply Mathlib/NumberTheory/LegendreSymbol/JacobiSymbol.lean:514]: I don't know enough about metaprogramming in Lean to deduplicate the code (when I did a naive replacement I ran into a deep recursion error). I suspect the usual strategy is that there is a split between the computation rule and norm_num extension which additionally focuses on producing small proof terms. For example there is also a lot of duplicate logic between Nat.fastFib and NormNum/NatFib.

**Comment by hanwenzhu** [comment]
I added the algorithm description in the module docstring

**Comment by urkud** [comment]
@digama0 Could you please have a look at the meta code?

**Review by digama0 — APPROVED** [review]

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by Ruben-VandeVelde** [comment]
This failed on bors. I merged master; hopefully this will make the error show up here.

**Comment by hanwenzhu** [comment]
Thank you very much @digama0 and @Ruben-VandeVelde ! The branch should compile now.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**